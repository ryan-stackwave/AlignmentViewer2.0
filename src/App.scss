/** App.scss */
.settings_box {
  font-size: 11px;
  margin: auto;
  width: 450px;
  border: 2px solid black;
  padding: 15px;
}
.settings_box h3 {
  margin-bottom: 0;
  text-align: center;
}
.settings_box h4 {
  margin-top: 0;
  text-align: center;
}
.settings_box h2 {
  padding-top: 0;
  margin-top: 0;
  text-align: center;
}

.settings_box .settings-header {
  position: relative;
}
.settings_box .settings-header button {
  position: absolute;
  right: 0;
  top: 0;
}

.settings_box label strong {
  width: 200px;
  display: inline-block;
}
.settings_box label select,
.settings_box label .zoom-level {
  width: 200px;
  display: inline-block;
}

.examples button.button-link {
  border: 1px solid transparent;
}

button.button-link {
  background: none !important;
  border: none;
  //padding: 0 !important;
  /*optional*/
  //font-family: arial, sans-serif;
  /*input has OS specific font-family*/
  color: #069;
  text-decoration: underline;
  cursor: pointer;
}

.hide {
  display: none;
}

.av_holder {
  padding-top: 20px;
}

.loader {
  position: absolute;
  bottom: 0;
  right: 0;
  border: 3px solid #f3f3f3; /* Light grey */
  border-top: 3px solid #3498db; /* Blue */
  border-radius: 50%;
  width: 14px;
  height: 14px;
  animation: spin 2s linear infinite;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
