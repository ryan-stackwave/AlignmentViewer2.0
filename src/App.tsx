import React from "react";
import "./App.scss";
import Alignment, { SequenceSortOptions } from "./common/Alignment";
import { AlignmentViewer } from "./components/AlignmentViewerComponent";
import {
  AminoAcidAlignmentStyle,
  NucleotideAlignmentStyle,
  AlignmentStyle,
  AlignmentTypes,
  PositionsToStyle,
  IColorScheme,
  ResidueDetailTypes,
} from "./common/MolecularStyles";
import { LOGO_TYPES } from "./components/SequenceLogoComponent";
import { FileInputComponent } from "./components/FileInputComponent";

export interface AppProps {}
export interface AppState {
  alignment?: Alignment;
  style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle;
  sortBy: SequenceSortOptions;
  logoPlotStyle: LOGO_TYPES;
  zoomLevel: number;
  showMiniMap: boolean;
  showAnnotations: boolean;
  showSettings: boolean;
  loading?: boolean;
}

export default class App extends React.Component<AppProps, AppState> {
  constructor(props: AppProps) {
    super(props);
    this.state = {
      style: new AminoAcidAlignmentStyle(),
      logoPlotStyle: LOGO_TYPES.LETTERS, //TODO - decide NT or AA based on alignment
      zoomLevel: 14,
      sortBy: SequenceSortOptions.INPUT,
      showMiniMap: false,
      showAnnotations: true,
      showSettings: true,
    };
  }

  async componentDidMount() {
    /*this.setState({
      alignment: await this.getAlignmentForFile(
        "7fa1c5691376beab198788a726917d48_b0.4.a2m"
      ),
      style: new AminoAcidAlignmentStyle()
    });*/
  }

  render() {
    const {
      alignment,
      logoPlotStyle,
      showAnnotations,
      showMiniMap,
      sortBy,
      style,
      zoomLevel,
    } = this.state;

    const alignmentElement = !alignment ? (
      <></>
    ) : (
      <div className="av_holder">
        <AlignmentViewer
          alignment={alignment}
          style={style}
          logoPlotStyle={logoPlotStyle}
          zoomLevel={zoomLevel}
          sortBy={sortBy}
          showMiniMap={showMiniMap}
          showAnnotations={showAnnotations}
        ></AlignmentViewer>
      </div>
    );

    return (
      <div>
        {this.renderSettingsBox(style)}
        {alignmentElement}
      </div>
    );
  }

  /*
  protected getAlignmentForFile = async (filename: string) => {
    const result = await fetch(`${process.env.PUBLIC_URL}/${filename}`);

    return Alignment.fromFileContents(filename, await result.text());
  };*/

  protected renderSettingsBox = (
    style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle
  ) => {
    const { alignment } = this.state;
    const alignmentDescription = alignment ? (
      <div>
        <h3>{alignment.getName()}</h3>
        <h4>{`${
          alignment.getSequences().length
        } sequences (rows) and ${alignment.getMaxSequenceLength()} positions (columns) `}</h4>
      </div>
    ) : (
      <></>
    );

    return (
      <div className="settings_box">
        <form>
          <div className="settings-header">
            <button
              style={{
                display: this.state.showSettings ? "none" : "block",
              }}
              type="button"
              className="button-link"
              onClick={(e) => {
                this.setState({
                  showSettings: true,
                });
              }}
            >
              Expand
            </button>
            <button
              style={{
                display: this.state.showSettings ? "block" : "none",
              }}
              type="button"
              className="button-link"
              onClick={(e) => {
                this.setState({
                  showSettings: false,
                });
              }}
            >
              Hide
            </button>
            <div>
              <h2>{`AlignmentViewer 2.0 Settings Demo`}</h2>
              {alignmentDescription}
            </div>
          </div>
          <div
            style={{
              display: this.state.showSettings ? "block" : "none",
              position: "relative",
            }}
          >
            {this.renderAlignmentTypeLabel(style)}
            {this.renderSortControl()}
            {this.renderColorScheme(style)}
            {this.renderResidueDetail(style)}
            {this.renderPositionStyling(style)}
            {this.renderSequenceLogo()}
            {this.renderZoomButtons()}
            {this.renderMiniMapToggle()}
            {this.renderAnnotationToggle()}
            <br></br>
            {this.renderFileUpload()}
            {this.renderExampleLinks()}
            <div
              className="loader"
              style={{
                display: this.state.loading ? "block" : "none",
              }}
            ></div>
          </div>
        </form>
      </div>
    );
  };

  protected renderSortControl = () => {
    const { sortBy } = this.state;
    return (
      <div>
        <label>
          <strong>Sort order:</strong>
          <select
            value={sortBy.key}
            onChange={(e) =>
              this.setState({
                sortBy: SequenceSortOptions.fromKey(e.target.value)!,
              })
            }
          >
            {SequenceSortOptions.list.map((sso) => {
              return (
                <option value={sso.key} key={sso.key}>
                  {sso.description}
                </option>
              );
            })}
          </select>
        </label>
      </div>
    );
  };
  protected renderAlignmentTypeLabel = (
    style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle
  ) => {
    return (
      <div>
        <label>
          <strong>Alignment Type:</strong>
          <select
            value={style.alignmentType.key}
            onChange={(e) =>
              this.setState({
                style: AlignmentStyle.fromAlignmentType(
                  AlignmentTypes.fromKey(e.target.value)!
                ),
              })
            }
          >
            {AlignmentTypes.list.map((alignmentType) => {
              return (
                <option value={alignmentType.key} key={alignmentType.key}>
                  {alignmentType.description}
                </option>
              );
            })}
          </select>
        </label>
      </div>
    );
  };
  protected renderColorScheme = (
    style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle
  ) => {
    return (
      <div>
        <label>
          <strong>Style:</strong>
          <select
            value={style.alignmentType.allColorSchemes.indexOf(
              style.colorScheme
            )}
            onChange={(e) => {
              this.setState({
                style: {
                  ...style!,
                  colorScheme:
                    style.alignmentType.allColorSchemes[
                      parseInt(e.target.value)
                    ],
                },
              });
            }}
          >
            {style.alignmentType.allColorSchemes.map(
              (colorScheme: IColorScheme, index: number) => {
                return (
                  <option key={index} value={index}>
                    {colorScheme.description}
                  </option>
                );
              }
            )}
          </select>
        </label>
      </div>
    );
  };
  protected renderResidueDetail = (
    style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle
  ) => {
    return (
      <div>
        <label>
          <strong>Residue Detail:</strong>
          <select
            value={style.residueDetail.key}
            onChange={(e) => {
              this.setState({
                style: {
                  ...style,
                  residueDetail: ResidueDetailTypes.fromKey(e.target.value)!,
                },
              });
            }}
          >
            {ResidueDetailTypes.list.map((rd) => {
              return (
                <option key={rd.key} value={rd.key}>
                  {rd.description}
                </option>
              );
            })}
          </select>
        </label>
      </div>
    );
  };
  protected renderPositionStyling = (
    style: AminoAcidAlignmentStyle | NucleotideAlignmentStyle
  ) => {
    return (
      <div>
        <label>
          <strong>Positions to Style:</strong>
          <select
            value={PositionsToStyle.list.indexOf(style.positionsToStyle)}
            onChange={(e) => {
              this.setState({
                style: {
                  ...style,
                  positionsToStyle:
                    PositionsToStyle.list[parseInt(e.target.value)],
                },
              });
            }}
          >
            {PositionsToStyle.list.map(
              (pts: PositionsToStyle, index: number) => {
                return (
                  <option key={index} value={index}>
                    {pts.description}
                  </option>
                );
              }
            )}
          </select>
        </label>
      </div>
    );
  };
  protected renderSequenceLogo = () => {
    const { logoPlotStyle } = this.state;
    return (
      <div>
        <label>
          <strong>Sequence Logo Style:</strong>
          <select
            value={logoPlotStyle}
            onChange={(e) => {
              this.setState({
                logoPlotStyle: e.target.value as LOGO_TYPES,
              });
            }}
          >
            {Object.values(LOGO_TYPES).map((logoType) => {
              return (
                <option key={logoType} value={logoType}>
                  {logoType}
                </option>
              );
            })}
          </select>
        </label>
      </div>
    );
  };
  protected renderZoomButtons = () => {
    const { zoomLevel } = this.state;
    return (
      <div>
        <label>
          <strong>Character size:</strong>
          <div className="zoom-level">
            <button
              type="button"
              disabled={zoomLevel < 3}
              onClick={(e) => {
                this.setState({
                  zoomLevel: zoomLevel - 1,
                });
              }}
            >
              -
            </button>
            {zoomLevel}
            <button
              type="button"
              disabled={zoomLevel > 15}
              onClick={(e) => {
                this.setState({
                  zoomLevel: zoomLevel + 1,
                });
              }}
            >
              +
            </button>
          </div>
        </label>
      </div>
    );
  };

  protected renderFileUpload = () => {
    return (
      <div>
        <FileInputComponent
          labelText={"Upload Sequence File:"}
          onFileLoadCb={this.onFileUpload}
        />
      </div>
    );
  };

  protected renderExampleLinks = () => {
    return (
      <div className="examples">
        <label>
          <strong>Example Alignments:</strong>
          <button
            type="button"
            className="button-link"
            onClick={async (e) => {
              this.setState({
                loading: true,
              });
              const f = new File(
                [
                  await (
                    await fetch(
                      `${process.env.PUBLIC_URL}/7fa1c5691376beab198788a726917d48_b0.4.a2m`
                    )
                  ).blob(),
                ],
                "beta_lactamase_example.fasta"
              );
              this.onFileUpload(f);
            }}
          >
            β-lactamase
          </button>
        </label>
      </div>
    );
  };

  protected renderMiniMapToggle = () => {
    return (
      <div className="minimap-toggle">
        <label>
          <strong>Show MiniMap:</strong>

          <input
            name="showMiniMap"
            type="checkbox"
            checked={this.state.showMiniMap}
            onChange={(e) => {
              const target = e.target;
              this.setState({
                showMiniMap: target.checked,
              });
            }}
          />
        </label>
      </div>
    );
  };

  protected renderAnnotationToggle = () => {
    return (
      <div className="annotation-toggle">
        <label>
          <strong>Show Annotations:</strong>

          <input
            name="showAnnotations"
            type="checkbox"
            checked={this.state.showAnnotations}
            onChange={(e) => {
              const target = e.target;
              this.setState({
                showAnnotations: target.checked,
              });
            }}
          />
        </label>
      </div>
    );
  };

  protected onFileUpload = async (file: File) => {
    //const fileText = await file.text(); //doesn't work in modern safari or even slightly older firefox
    this.setState({
      loading: true,
    });
    var reader = new FileReader();
    reader.onload = (e) => {
      const fileText = reader.result as string;
      const alignment = Alignment.fromFileContents(file.name, fileText);
      this.setState({
        alignment: alignment,
        loading: false,
      });
    };
    reader.readAsText(file);
  };
}
