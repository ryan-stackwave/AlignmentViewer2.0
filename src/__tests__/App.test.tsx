import "jest-webgl-canvas-mock";
import * as fs from "fs";
import * as React from "react";

import { shallow, default as Enzyme } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
Enzyme.configure({ adapter: new Adapter() });

import { enableFetchMocks, default as fetchMock } from "jest-fetch-mock";
enableFetchMocks();

import { default as App } from "../App";

// Due to the runtime necessities of using styles, we need to explicitly mock out some stub data.
// https://github.com/facebook/jest/issues/3094
jest.mock("../MolecularStyles.module.scss", () => {
  return {
    aaStyBGAlpha_Default: "",
    aaStyClass_Default: "",
    aaStyColorOrder_Default: "",
    aaStyColors_Default: "",
    aaStyDesc_Default: "",
    ntStyBGAlpha_Default: "",
    ntStyClass_Default: "",
    ntStyColorOrder_Default: "",
    ntStyColors_Default: "",
    ntStyDesc_Default: ""
  };
});

describe("App", () => {
  beforeAll(() => {
    const sequenceFile = fs
      .readFileSync("public/7fa1c5691376beab198788a726917d48_b0.4.a2m")
      .toString();
    fetchMock.mockResponse(sequenceFile);
  });

  it("Should render within 2 minutes.", async () => {
    const wrapper = shallow(<App />);
    wrapper.update();
    await new Promise(resolve => setImmediate(resolve));
    expect(wrapper).toMatchSnapshot();
  }, 120000);
});
