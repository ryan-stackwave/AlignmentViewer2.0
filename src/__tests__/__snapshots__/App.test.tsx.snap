// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should render within 2 minutes. 1`] = `
<div>
  <div
    className="settings_box"
  >
    <form>
      <div
        className="settings-header"
      >
        <button
          className="button-link"
          onClick={[Function]}
          style={
            Object {
              "display": "none",
            }
          }
          type="button"
        >
          Expand
        </button>
        <button
          className="button-link"
          onClick={[Function]}
          style={
            Object {
              "display": "block",
            }
          }
          type="button"
        >
          Hide
        </button>
        <div>
          <h2>
            AlignmentViewer 2.0 Settings Demo
          </h2>
        </div>
      </div>
      <div
        style={
          Object {
            "display": "block",
            "position": "relative",
          }
        }
      >
        <div>
          <label>
            <strong>
              Alignment Type:
            </strong>
            <select
              onChange={[Function]}
              value="aminoacid-alignment"
            >
              <option
                key="aminoacid-alignment"
                value="aminoacid-alignment"
              >
                Amino Acid Sequences
              </option>
              <option
                key="nucleotide-alignment"
                value="nucleotide-alignment"
              >
                Nucleotide Sequences
              </option>
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Sort order:
            </strong>
            <select
              onChange={[Function]}
              value="as-input"
            >
              <option
                key="as-input"
                value="as-input"
              >
                As input
              </option>
              <option
                key="hamming-dist-to-target"
                value="hamming-dist-to-target"
              >
                Hamming distance to target sequence
              </option>
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Style:
            </strong>
            <select
              onChange={[Function]}
              value={0}
            >
              <option
                key="0"
                value={0}
              />
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Residue Detail:
            </strong>
            <select
              onChange={[Function]}
              value="both"
            >
              <option
                key="both"
                value="both"
              >
                Style Background and Letters
              </option>
              <option
                key="letter-only"
                value="letter-only"
              >
                Only Style Letters
              </option>
              <option
                key="background-only"
                value="background-only"
              >
                Only Style Background
              </option>
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Positions to Style:
            </strong>
            <select
              onChange={[Function]}
              value={0}
            >
              <option
                key="0"
                value={0}
              >
                All
              </option>
              <option
                key="1"
                value={1}
              >
                Same as Query
              </option>
              <option
                key="2"
                value={2}
              >
                Different from Query
              </option>
              <option
                key="3"
                value={3}
              >
                Same as Consensus
              </option>
              <option
                key="4"
                value={4}
              >
                Different from Consensus
              </option>
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Sequence Logo Style:
            </strong>
            <select
              onChange={[Function]}
              value="letter stack"
            >
              <option
                key="letter stack"
                value="letter stack"
              >
                letter stack
              </option>
              <option
                key="bar plot"
                value="bar plot"
              >
                bar plot
              </option>
            </select>
          </label>
        </div>
        <div>
          <label>
            <strong>
              Character size:
            </strong>
            <div
              className="zoom-level"
            >
              <button
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                -
              </button>
              14
              <button
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                +
              </button>
            </div>
          </label>
        </div>
        <div
          className="minimap-toggle"
        >
          <label>
            <strong>
              Show MiniMap:
            </strong>
            <input
              checked={false}
              name="showMiniMap"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </div>
        <div
          className="annotation-toggle"
        >
          <label>
            <strong>
              Show Annotations:
            </strong>
            <input
              checked={true}
              name="showAnnotations"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </div>
        <br />
        <div>
          <FileInputComponent
            labelText="Upload Sequence File:"
            onFileLoadCb={[Function]}
          />
        </div>
        <div
          className="examples"
        >
          <label>
            <strong>
              Example Alignments:
            </strong>
            <button
              className="button-link"
              onClick={[Function]}
              type="button"
            >
              Î²-lactamase
            </button>
          </label>
        </div>
        <div
          className="loader"
          style={
            Object {
              "display": "none",
            }
          }
        />
      </div>
    </form>
  </div>
</div>
`;
